pilot/pkg/dataplane/interfaces.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dataplane
+
+import (
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2"
+
+	"istio.io/istio/pilot/pkg/model"
+)
+
+// Dataplane represents the interfaces to be implemented by code that generates xDS responses
+type Dataplane interface {
+	// BuildListeners returns the list of listeners for the given proxy. This is the LDS output
+	BuildListeners(env model.Environment, node model.Proxy) []*v2.Listener
+
+	// BuildClusters returns the list of clusters for the given proxy. This is the CDS output
+	BuildClusters(env model.Environment, node model.Proxy) []*v2.Cluster
+
+	// BuildRoutes returns the list of routes for the given proxy. This is the RDS output
+	BuildRoutes(env model.Environment, node model.Proxy, routeName string) []*v2.RouteConfiguration
+}
+
+// PluginCallbacks represents the interfaces implemented by code that modifies the default output of
+// networking. Examples include AuthenticationPlugin that sets up mTLS authentication on the inbound Listener
+// and outbound Cluster, the mixer plugin that sets up policy checks on the inbound listener, etc.
+// TODO: determine the function params after looking at use cases
+type PluginCallbacks interface {
+	// OnOutboundListener is called whenever a new outbound listener is added to the LDS output for a given service
+	// Can be used to add additional filters on the outbound path
+	OnOutboundListener(env model.Environment, node model.Proxy, service *model.Service, listener *v2.Listener) *v2.Listener
```

Comment:
This isn't right, as we want to model outbound routes using filter chains, not listeners.

Reply:
I am going by what the current code looks like. This plugin interface is not published. So we can iterate on this as we see fit. But the primary target is to enable simpler testing, and cleaner code separation.

Reply:
e.g. `use_original_dst_port` is on deprecation notice

Reply:
As long as this interface is marked as temporary or deprecated, I'm fine.